mqtt_subscriber/oli_broker.py

```diff
@@ -3,40 +3,64 @@
 import logging
 import traceback
 import paho.mqtt.client as mqtt
+from d3a_api_client.rest_device import RestDeviceClient
+from mqtt_subscriber import allowed_devices_name_mapping, generate_topic_api_client_args_mapping
+from time import time
+
+
+class MQTTBrokerConnectionError(Exception):
+    pass
 
 
 MQTT_PORT = 1883
 MINUTES_IN_HOUR = 60
 MEASUREMENT_PERIOD_MINUTES = 15
+RELOAD_CN_DEVICE_LIST_TIMEOUT_SECONDS = 60 * MEASUREMENT_PERIOD_MINUTES
 
 
 class MQTTConnection:
-    def __init__(self, api_client_dict):
-        self.api_clients = api_client_dict
+    def __init__(self, topic_api_client_dict):
+        self.topic_api_client_dict = topic_api_client_dict
         self.client = mqtt.Client()
         self.client.on_connect = self.on_connect
         self.client.on_message = self.on_message
+        self.last_time_checked = time()
 
     def on_connect(self, client, userdata, flags, rc):
         logging.info(f"Connected with result code {str(rc)}")
 
-        for device_name in self.api_clients:
-            client.subscribe(f"{device_name}/activeEnergy/#")
+        for topic_name in allowed_devices_name_mapping.values():
+            logging.info(f"Subscribed to topic {topic_name}")
+            subscribe_result = client.subscribe(topic_name)
+            if subscribe_result[0] != 0:
+                raise MQTTBrokerConnectionError(
+                    f"Could not subscribe to topic {topic_name}. Subscribe error code {subscribe_result}."
+                )
 
     def on_message(self, client, userdata, msg):
         logging.info(f"Received mqtt message {msg.topic} {str(msg.payload)}")
         try:
+            self.refresh_cn_and_device_list()
+
             payload = json.loads(msg.payload.decode("utf-8"))
 
-            device_name = msg.topic.split("/activeEnergy/")[0]
             energy = payload["value"]
             power = energy * (MINUTES_IN_HOUR / MEASUREMENT_PERIOD_MINUTES)
-            self.api_clients[device_name].set_power_forecast(power)
+
+            # Transmit power values to CN
+            for api_args in self.topic_api_client_dict[msg.topic]:
+                RestDeviceClient(**api_args).set_power_forecast(power, do_not_wait=True)
```

Comment:
So you are re-initiating the RestDevices every market slot?

Reply:
Yes, for simplicity's sake I thought it was better to at first, completely ignore the response from the d3a via the websocket and just send the request. That way we do not have to maintain long-running connections that are prone to crash due to e.g. a network error.